trainer/lightgbm_gridsearch.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2018 Johanna de Vos
+# Copyright 2018 Tom Westerhout
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import, division, print_function
+import json
+import logging
+from os import path
+
+import lightgbm as lgb
+from sklearn.model_selection import GridSearchCV
+
+import trainer.lightgbm_functions as lf
+import trainer.preprocessing as pp
+
+
+# Default parameters
+LGBM_PARAMS = {
+    'boosting_type':      'gbdt',
+    'objective':          'binary',
+    'metric':             'auc',
+    'learning_rate':      0.08,
+    'num_leaves':         31,  # we should let it be smaller than 2^(max_depth)
+    'max_depth':          -1,  # -1 means no limit
+    'min_child_samples':  20,  # Minimum number of data need in a child(min_data_in_leaf)
+    'max_bin':            255,  # Number of bucketed bin for feature values
+    'subsample':          0.6,  # Subsample ratio of the training instance.
+    'subsample_freq':     0,  # frequence of subsample, <=0 means no enable
+    'colsample_bytree':   0.3,  # Subsample ratio of columns when constructing each tree.
+    'min_child_weight':   5,  # Minimum sum of instance weight(hessian) needed in a child(leaf)
+    'subsample_for_bin':  200000,  # Number of samples for constructing bin
+    'min_split_gain':     0,  # lambda_l1, lambda_l2 and min_gain_to_split to regularization
+    'reg_alpha':          0,  # L1 regularization term on weights
+    'reg_lambda':         0,  # L2 regularization term on weights
+    'nthread':            8,
+    'verbose':            0,
+}
+
+
+# Parameters to be optimized
+LGBM_PARAM_GRID = {
+    'learning_rate':      [0.05, 0.08], # NB: Use 'range' or something similar
+    'num_leaves':         [30, 31],  # we should let it be smaller than 2^(max_depth)
+}
+
+
+
+def lgb_gridsearch(default_params, param_grid, training_data, predictors, 
+                   target, validation_data=None, categorical_features=None, 
+                   n_splits=2, early_stopping_rounds=20):
+    """
+    Performs a grid search to find the optimal value for all parameters.
+    The grid search makes use of k-fold cross-validation.
+    Returns a dictionary with the optimal value for all parameters."
+
+    """
+    
+    # Instantiate classification model with the default parameter values
+    gbm = lgb.LGBMRegressor(**default_params)
+    
+    # Dictionary with additional parameters to pass to .fit method
+    fit_params = {
```

Comment:
At this point fit_params are never actually passed to the fit method if I am not mistaken. Why so?

Reply:
Good point. There is not real reason, I think I just forgot. Perhaps this is the explanation for the difference between gridsearch and main?

Reply:
I ran the gridsearch having added 'fit_params', and I got both a deprecation warning and much worse results:

`C:\Users\johan\Anaconda3\envs\python27\lib\site-packages\sklearn\model_selection\_search.py:584: DeprecationWarning: "fit_params" as a constructor argument was deprecated in version 0.19 and will be removed in version 0.21. Pass fit parameters to the "fit" method instead.
  '"fit" method instead.', DeprecationWarning)
Fitting 5 folds for each of 4 candidates, totalling 20 fits
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    8.1s finished
[2018-05-08 10:44:36,936] [INFO] The best score from the grid search was: 0.780355036233
[2018-05-08 10:44:36,940] [INFO] It was obtained with the following parameters: {'num_leaves': 30, 'learning_rate': 0.08}`

When running main without `fit_params` the result is suddenly 0.948. So suddenly is happening there. Well-spotted, @aranas :)